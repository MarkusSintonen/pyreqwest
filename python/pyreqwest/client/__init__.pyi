# ruff: noqa: E501, F401
from datetime import timedelta
from types import TracebackType
from typing import Mapping
from typing_extensions import Self

from pyreqwest.http import UrlType
from pyreqwest.middleware import Middleware
from pyreqwest.proxy import Proxy
from pyreqwest.request import RequestBuilder


class Client:
    def request(self, method: str, url: UrlType) -> RequestBuilder: ...
    def get(self, url: UrlType) -> RequestBuilder: ...
    def post(self, url: UrlType) -> RequestBuilder: ...
    def put(self, url: UrlType) -> RequestBuilder: ...
    def patch(self, url: UrlType) -> RequestBuilder: ...
    def delete(self, url: UrlType) -> RequestBuilder: ...
    def head(self, url: UrlType) -> RequestBuilder: ...
    async def __aenter__(self) -> Self: ...
    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None: ...
    async def close(self) -> None: ...


class ClientBuilder:
    def __init__(self) -> None: ...
    def build(self) -> Client: ...
    def with_middleware(self, middleware: Middleware) -> Self: ...
    def max_connections(self, max: int) -> Self: ...
    def error_for_status(self, enable: bool) -> Self: ...
    def user_agent(self, value: str) -> Self: ...
    def default_headers(self, headers: Mapping[str, str]) -> Self: ...
    def cookie_store(self, enable: bool) -> Self: ...
    def gzip(self, enable: bool) -> Self: ...
    def brotli(self, enable: bool) -> Self: ...
    def zstd(self, enable: bool) -> Self: ...
    def deflate(self, enable: bool) -> Self: ...
    def max_redirects(self, max: int) -> Self: ...
    def referer(self, enable: bool) -> Self: ...
    def proxy(self, proxy: Proxy) -> Self: ...
    def no_proxy(self) -> Self: ...
    def timeout(self, timeout: timedelta) -> Self: ...
    def read_timeout(self, timeout: timedelta) -> Self: ...
    def connect_timeout(self, timeout: timedelta) -> Self: ...
    def connection_verbose(self, verbose: bool) -> Self: ...
    def pool_idle_timeout(self, timeout: timedelta) -> Self: ...
    def pool_max_idle_per_host(self, max: int) -> Self: ...
    def http1_title_case_headers(self) -> Self: ...
    def http1_allow_obsolete_multiline_headers_in_responses(self, value: bool) -> Self: ...
    def http1_ignore_invalid_headers_in_responses(self, value: bool) -> Self: ...
    def http1_allow_spaces_after_header_name_in_responses(self, value: bool) -> Self: ...
    def http1_only(self) -> Self: ...
    def http09_responses(self) -> Self: ...
    def http2_prior_knowledge(self) -> Self: ...
    def http2_initial_stream_window_size(self, value: int | None) -> Self: ...
    def http2_initial_connection_window_size(self, value: int | None) -> Self: ...
    def http2_adaptive_window(self, enabled: bool) -> Self: ...
    def http2_max_frame_size(self, value: int | None) -> Self: ...
    def http2_max_header_list_size(self, value: int) -> Self: ...
    def http2_keep_alive_interval(self, value: timedelta | None) -> Self: ...
    def http2_keep_alive_timeout(self, timeout: timedelta) -> Self: ...
    def http2_keep_alive_while_idle(self, enabled: bool) -> Self: ...
    def tcp_nodelay(self, enabled: bool) -> Self: ...
    def local_address(self, addr: str | None) -> Self: ...
    def interface(self, value: str) -> Self: ...
    def tcp_keepalive(self, value: timedelta | None) -> Self: ...
    def tcp_keepalive_interval(self, value: timedelta | None) -> Self: ...
    def tcp_keepalive_retries(self, value: int | None) -> Self: ...
    def tcp_user_timeout(self, timeout: timedelta | None) -> Self: ...
    def add_root_certificate_der(self, cert: bytes) -> Self: ...
    def add_root_certificate_pem(self, cert: bytes) -> Self: ...
    def add_crl_pem(self, cert: bytes) -> Self: ...
    def tls_built_in_root_certs(self, enable: bool) -> Self: ...
    def tls_built_in_webpki_certs(self, enable: bool) -> Self: ...
    def identity_pem(self, buf: bytes) -> Self: ...
    def danger_accept_invalid_hostnames(self, enable: bool) -> Self: ...
    def danger_accept_invalid_certs(self, enable: bool) -> Self: ...
    def tls_sni(self, enable: bool) -> Self: ...
    def min_tls_version(self, value: str) -> Self: ...
    def max_tls_version(self, value: str) -> Self: ...
    def tls_info(self, enable: bool) -> Self: ...
    def https_only(self, enable: bool) -> Self: ...
    def resolve(self, domain: str, ip: str, port: int) -> Self: ...
