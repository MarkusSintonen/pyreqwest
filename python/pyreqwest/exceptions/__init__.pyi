class RequestError(Exception): ...

class SendError(RequestError): ...
class SendConnectionError(SendError): ...
class SendBodyError(SendError): ...
class SendTimeoutError(SendError): ...
class PoolTimeoutError(SendTimeoutError): ...

class ReadError(RequestError): ...
class ReadBodyError(ReadError): ...
class ReadTimeoutError(ReadError): ...
