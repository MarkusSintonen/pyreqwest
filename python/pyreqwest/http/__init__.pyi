from typing import Any


class Url:
    def __init__(self, url: str) -> None: ...
    @staticmethod
    def parse(url: str) -> 'Url': ...
    @staticmethod
    def parse_with_params(url: str, params: list[tuple[str, str]]) -> 'Url': ...

    def join(self, input: str) -> 'Url': ...
    def make_relative(self, base: 'Url') -> str | None: ...
    def as_str(self) -> str: ...
    def origin_ascii(self) -> str: ...
    def origin_unicode(self) -> str: ...
    def scheme(self) -> str: ...
    def is_special(self) -> bool: ...
    def has_authority(self) -> bool: ...
    def authority(self) -> str: ...
    def cannot_be_a_base(self) -> bool: ...
    def username(self) -> str: ...
    def password(self) -> str | None: ...
    def has_host(self) -> bool: ...
    def host_str(self) -> str | None: ...
    def domain(self) -> str | None: ...
    def port(self) -> int | None: ...
    def port_or_known_default(self) -> int | None: ...
    def path(self) -> str: ...
    def path_segments(self) -> list[str] | None: ...
    def query(self) -> str | None: ...
    def query_pairs(self) -> list[tuple[str, str]]: ...
    def fragment(self) -> str | None: ...
    def set_fragment(self, fragment: str | None) -> None: ...
    def set_query(self, query: str | None) -> None: ...
    def set_query_pairs(self, query: list[tuple[str, str]]) -> None: ...
    def set_path(self, path: str) -> None: ...
    def set_path_segments(self, segments: list[str]) -> None: ...
    def set_port(self, port: int | None) -> None: ...
    def set_host(self, host: str | None) -> None: ...
    def set_ip_host(self, addr: str) -> None: ...
    def set_password(self, password: str | None) -> None: ...
    def set_username(self, username: str) -> None: ...
    def set_scheme(self, scheme: str) -> None: ...
    def __repr__(self) -> str: ...
    def __str__(self) -> str: ...
    def __hash__(self) -> int: ...
    def __richcmp__(self, other: Any, op: Any) -> bool: ...


UrlType = Url | str
