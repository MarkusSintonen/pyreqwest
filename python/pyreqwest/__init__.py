"""pyreqwest - Powerful and fast Rust based HTTP client library.

Inspired and using [reqwest](https://github.com/seanmonstar/reqwest).

Feature-rich:
- High performance
- Asynchronous and synchronous HTTP clients
- Built-in support for JSON, form data, and multipart requests
- Connection pooling and keep-alive
- Timeout and retry policies
- HTTPS/TLS support (using [rustls](https://github.com/rustls/rustls))
- Middleware support for advanced request/response processing
- Streaming request and response bodies
- Easy integration with web frameworks
- Cookie management
- Mocking and testing utilities
- Type-safe APIs with Python type hints
- Free threading
"""
