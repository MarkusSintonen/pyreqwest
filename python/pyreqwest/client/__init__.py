from pyreqwest._pyreqwest.client import *
