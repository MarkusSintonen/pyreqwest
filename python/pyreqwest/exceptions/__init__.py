from json import JSONDecodeError as JSONDecodeError_
from typing import TypedDict, Generic, TypeVar, Any

Cause = TypedDict("Cause", {"message": str})
CauseErrorDetails = TypedDict("CauseErrorDetails", {"causes": list[Cause] | None})
StatusErrorDetails = TypedDict("StatusErrorDetails", {"status": int})
T = TypeVar("T", bound=CauseErrorDetails | StatusErrorDetails)


class HTTPError(Exception):
    def __init__(self, message: str, *args: Any) -> None:
        assert isinstance(message, str)
        super().__init__(message, *args)
        self.message = message

class DetailedHTTPError(HTTPError, Generic[T]):
    def __init__(self, message: str, details: T | None = None) -> None:
        assert isinstance(details, dict | None)
        super().__init__(message, details)
        self.details = details

class RequestError(DetailedHTTPError[T], Generic[T]): ...

class StatusError(RequestError[StatusErrorDetails]): ...
class RedirectError(RequestError[CauseErrorDetails]): ...
class RequestPanicError(RequestError[CauseErrorDetails]): ...
class BodyError(RequestError[CauseErrorDetails]): ...
class TransportError(RequestError[CauseErrorDetails]): ...

class DecodeError(BodyError): ...

class RequestTimeoutError(TransportError, TimeoutError): ...
class NetworkError(TransportError): ...

class ConnectTimeoutError(RequestTimeoutError): ...
class ReadTimeoutError(RequestTimeoutError): ...
class WriteTimeoutError(RequestTimeoutError): ...
class PoolTimeoutError(RequestTimeoutError): ...

class ConnectError(NetworkError): ...
class ReadError(NetworkError): ...
class WriteError(NetworkError): ...
class CloseError(NetworkError): ...

class BuilderError(DetailedHTTPError[CauseErrorDetails], ValueError): ...

class JSONDecodeError(HTTPError, JSONDecodeError_):
    def __init__(self, message: str, details: dict[str, Any]) -> None:
        assert isinstance(details, dict)
        assert isinstance(details["doc"], str) and isinstance(details["pos"], int)
        super().__init__(message, details["doc"], details["pos"])
