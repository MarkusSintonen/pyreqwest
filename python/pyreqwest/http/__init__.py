from pyreqwest._pyreqwest.http import *
