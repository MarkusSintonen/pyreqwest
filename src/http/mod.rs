mod body;
mod header_map;
mod mime;
mod types;
mod url;

pub use body::*;
pub use header_map::*;
pub use mime::*;
pub use types::*;
pub use url::*;
