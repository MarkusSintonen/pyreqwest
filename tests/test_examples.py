import importlib
from pathlib import Path
from typing import Any

import pytest
from pyreqwest.http import Url
from syrupy import SnapshotAssertion  # type: ignore[attr-defined]

from examples._utils import run_examples

EXAMPLE_MODULES = [
    p.stem
    for p in (Path(__file__).parent.parent / "examples").iterdir()
    if p.suffix == ".py" and not p.name.startswith("_")
]


@pytest.mark.parametrize("example", EXAMPLE_MODULES)
async def test_examples(
    capsys: pytest.CaptureFixture[str],
    httpbin: Any,
    snapshot: SnapshotAssertion,
    monkeypatch: pytest.MonkeyPatch,
    example: str,
) -> None:
    url = Url(httpbin.url)
    monkeypatch.setenv("HTTPBIN", str(url))

    await run_examples(importlib.import_module(f"examples.{example}"))

    normalized = capsys.readouterr().out.replace(f":{url.port}/", ":<PORT>/")
    assert normalized == snapshot
