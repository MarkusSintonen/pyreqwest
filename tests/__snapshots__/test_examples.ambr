# serializer version: 1
# name: test_examples[basic_client]
  '''
  
  # running: example_simple_get
  {'args': {'q': 'pyreqwest'}, 'url': 'https://httpbin.org/get?q=pyreqwest', 'status': 200}
  
  # running: example_simple_get_sync
  {'args': {'q': 'pyreqwest'}, 'url': 'https://httpbin.org/get?q=pyreqwest'}
  
  # running: example_url_usage
  {'args': {'q': 'pyreqwest'}, 'url': 'https://httpbin.org/get?q=pyreqwest'}
  
  # running: example_error_for_status
  {'error': "('HTTP status client error', {'status': 400})"}
  {'status': 400}
  
  # running: example_base_url
  {'status': 200, 'body': 'foobar'}
  
  # running: example_read_bytes
  {'status': 200, 'body': Bytes(b"\xc5\xd7\x14\x84\xf8\xcf\x9b\xf4\xb7oG\x90G0\x80K")}
  
  # running: example_read_text
  {'status': 200, 'text': '<h1>Unicode Demo</h1>'}
  
  # running: example_headers
  {'x_client': 'client_value', 'x_req': 'req_value', 'x_req2': 'req2_value', 'user_agent': 'ua-example/1.0'}
  
  # running: example_stream_download
  {'chunks': 5, 'total_bytes': 500}
  
  # running: example_stream_upload
  {'status': 200, 'data': 'part-0_part-1_part-2_part-3_part-4_'}
  
  # running: example_timeouts
  {'error': "('connection timeout', {'causes': [{'message': 'error sending request for url (https://httpbin.org/delay/2)'}, {'message': 'operation timed out'}]})"}
  
  '''
# ---
# name: test_examples[json_usage]
  '''
  
  # running: example_post_json
  {'status': 200, 'echo': {'message': 'hello'}, 'content_type': 'application/json'}
  
  # running: example_custom_json_dumps
  {'status': 200, 'json': {'value': 'Zm9v'}}
  
  # running: example_custom_json_loads
  {'status': 200, 'json': {'value': b'foo'}}
  
  '''
# ---
# name: test_examples[middleware]
  '''
  
  # running: example_simple
  {'request_header': 'val1', 'response_header': 'val2'}
  
  # running: example_multiple
  {'order': ['1', '2']}
  
  # running: example_extensions
  {'mws': ['init', 'init_mw1'], 'extensions': {'demo_ext': 'init_mw1_mw2'}}
  
  # running: example_retry_middleware
  {'url': 'https://httpbin.org/delay/0', 'status': 200, 'retried': True}
  
  # running: example_middleware_modify_request_body
  {'data': 'hello_world'}
  
  # running: example_middleware_modify_request_body_streamed
  {'data': 'hello_world_middleware'}
  
  # running: example_middleware_override_response_builder
  {'status': 202, 'body': 'overridden'}
  
  # running: example_middleware_request_specific
  {'req1_args': {'mw1': 'val1', 'mw2': 'val2'}, 'req2_args': {'mw1': 'val1'}}
  
  '''
# ---
